### PR TITLE
Fix copying duplicate file name conflicts.

### PR DESCRIPTION
**Pull request by NickGoog**
Applies to mv logic as well. Basically treats destination as an existing
directory if multiple top-level sources are detected.

### CODE REVIEW

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:693]: This is no longer valid?

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:685]: Is this still specific to mv? I thought we were planning to change this behavior for cp as well?

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:713]: If we remove the mv condition from the above if condition, cand we drop this condition in that case. From the way I understand it now, the behavior for mv and cp is going to be the same. In that case, I think if we drop the perform_mv condition, do we still need this check?

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:699]: I wonder if we need the second condition after the OR here. preserve_src_top_level_dirs kinda guarantees that the destination is an existing dir correct? And I think we don't care if the source is a container as long as we have to preserve the top level dirs?

**Review by dilipped — COMMENTED** [review]
Yet to review the tests.

File: gslib/command.py
Comment (NickGoog) [reply gslib/command.py:899]: kwargs are kool

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:713]: The `mv` code in the "Case 1" block is:
```
if (global_copy_helper_opts.perform_mv and recursion_requested
    and src_url_names_container and not preserve_src_top_level_dirs):
  exp_src_url_tail = (exp_src_url.url_string[len(src_url.url_string):])
  dst_key_name = '%s/%s' % (exp_dst_url.object_name.rstrip('/'),
                              exp_src_url_tail.strip('/'))
```
To get to this block you're commenting on, you have to meet this condition (combining "Case 2" and `not preserve_src_top_level_dirs`):
```
if src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory():
  src_url_path_sans_final_dir = GetPathBeforeFinalDir(src_url, exp_src_url)
  dst_key_name = exp_src_url.versionless_url_string[
          len(src_url_path_sans_final_dir):].lstrip(src_url.delim)
  dst_key_name = dst_key_name.partition(src_url.delim)[-1]
```

It looks like there isn't an overlap of logic we can condense.

File: gslib/utils/copy_helper.py
Comment (dilipped) [reply gslib/utils/copy_helper.py:685]: Right, but what I am saying is that wouldn't the behavior for mv and cp end up being the same with the new naming rules? do we have to treat mv and cp differently?

Comment (NickGoog) [reply gslib/utils/copy_helper.py:685]: I don't think the behavior is the same. This block is more like "rename" behavior.
More details in responses below.

Comment (dilipped) [reply gslib/utils/copy_helper.py:685]: Discussed offline. Waiting for Nick to provide command examples/scenarios where mv and cp will have different behavior.

Comment (NickGoog) [reply gslib/utils/copy_helper.py]: Removing Case I looks like a success

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:673]: Do we need the case (src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory())) 

I am a bit confused here. Based on the description about for case 1, seems like we are handing only the preserve_src_top_level_dirs for case 1. This means a rename like operation `gsutil mv dir gs://non-existing_dir` will be handled by the else part correct? So do we need to have anything other than preserve_src_top_level_dirs here?

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:664]: Done

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:676]: Oops, outdated code comments

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:696]: Ditto

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:668]: Restructured the entire block

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:680]: I responded to the [earlier thread](https://github.com/GoogleCloudPlatform/gsutil/pull/1465#discussion_r813417927) about this line by explaining the boolean logic differences. Also note there is a `if not preserve_src_top_level_dirs` block for handling the case `(src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory()))`.

And tests fail without the case. Here's one example:
```
FAIL: testCopyingSubdirRecursiveToNonexistentSubdir (gslib.tests.test_naming.GsutilNamingTests)
Tests copying a directory with a single file recursively to a bucket.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/hartunian/Desktop/ide/gsutil/gslib/tests/test_naming.py", line 634, in testCopyingSubdirRecursiveToNonexistentSubdir
    self.assertEqual(expected, actual)
AssertionError: Items in the first set but not the second:
'gs://gsutil-test-testcopyingsubdirrecursivetononexisten-5c92d77f/dir3/dir2/foo'
Items in the second set but not the first:
'gs://gsutil-test-testcopyingsubdirrecursivetononexisten-5c92d77f/dir3/foo'
```

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:666]: I like the new comment!

File: gslib/utils/copy_helper.py
Comment (dilipped) [reply gslib/utils/copy_helper.py:680]: Got it. Thanks for checking

File: gslib/utils/copy_helper.py
Comment (dilipped) [inline gslib/utils/copy_helper.py:707]: Don't we have to handle special cases for `.` and `..` like we do in the `gcloud storage` code? Or is that handled already?

File: gslib/tests/test_cp.py
Comment (dilipped) [inline gslib/tests/test_cp.py:1720]: I think a similar test should be created for non-existing destination dir as well, because that's the key change that this PR is introducing.

File: gslib/tests/test_mv.py
Comment (dilipped) [inline gslib/tests/test_mv.py:135]: If I understand correctly, this would work fine even without the change made for multiple top level sources given that the destination exists. I think we should check for a non-existing destination dir.

File: gslib/tests/test_naming.py
Comment (dilipped) [inline gslib/tests/test_naming.py:984]: If you change this to `'f3'),])` (adding a comma after the tuple), I think yapf will format this in a slightly better way.

**Comment by NickGoog** [comment]
Removing the two commit structure to make merging in the requested changes simpler

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:707]: Thanks

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:666]: Thanks

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:135]: I'm not sure it's the case that this would work before the PR.
This also copies to a non-existing destination dir (it copies to bucket).
But will specifically add a test copying to a non-existing bucket subdir.

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:707]: Don't want to merge this PR with fixes for the `..` issue @rrauber is handling.

File: gslib/tests/test_cp.py
Comment (dilipped) [reply gslib/tests/test_cp.py]: bucket will get treated as an existing dir.

File: gslib/utils/copy_helper.py
Comment (dilipped) [reply gslib/utils/copy_helper.py:707]: That will lead to differences in the behavior for `..` upload case between gsutil and gcloud. Also, I think Ross is going to simply skip downloading any file that has `..` in the object name, which seems to be different than the upload case anyway.

File: gslib/tests/test_mv.py
Comment (dilipped) [reply gslib/tests/test_mv.py:135]: Bucket gets treated as an existing dir, since a bucket always exists if it is created.

File: gslib/utils/copy_helper.py
Comment (dilipped) [reply gslib/utils/copy_helper.py:707]: This is not addressed?

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:192]: Adds a lot of overhead and checking the expected number of result items has been sufficient for the other tests in this file.

File: gslib/tests/test_mv.py
Comment (dilipped) [inline gslib/tests/test_mv.py:156]: For this test, I think it's important to check the destination objects path. e.g how can we tell if the destination object is gs://bucket/dir/dir1/file.txt or gs://bucket/dir1/file1.txt

File: gslib/tests/test_mv.py
Comment (dilipped) [reply gslib/tests/test_mv.py:192]: Are you talking about the overhead in terms of code? In that case, you can have a helper method. But anyway, I have left a comment for test_move_bucket_objects_with_duplicate_names_to_bucket_subdir where such kind of checking is required. So if you create such a helper, you might be able to reuse it here.

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:156]: This is the focus of `test_naming.py`, specifically `testRecursiveCopyObjsAndFilesToNonExistentBucketSubDir`

Comment (dilipped) [reply gslib/tests/test_mv.py:156]: Agreed, but there is nothing in this test module that tells us that naming gets called. Plus, it's a less toll on whoever is not familiar with the behavior. Tests are great for documenting the behaviors and having the expected output checked will make it easier to understand how the mv behavior is meant to work for complicated naming behavior like these. Also, we check for the output for the cp tests anyway, why not do the same here?

Comment (NickGoog) [reply gslib/tests/test_mv.py:156]: >> but there is nothing in this test module that tells us that naming gets called

Not sure what you mean here. E2E tests are normally higher level than making sure underlying functions get called.

>> having the expected output checked will make it easier to understand how the mv behavior is meant to work

Sure, but I'd rather describe what's happening with a code comment than adding `ls` API calls and breaking the pattern of the rest of the file.

>> we check for the output for the cp tests anyway, why not do the same here

The `mv` tests are largely duplicates of functionality  tested in `test_cp.py`. The original authors probably meant for this to be lighter weight because (1) E2E tests are costly to run and (2) `test_cp.py` already does more detailed output analysis. 

I technically didn't need to add anything to `test_mv.py` because we already get  coverage from `test_cp.py`, but I wanted to be thorough. It's pushing it too far to break the lightweight testing pattern of the rest of the file, and I'd rather delete all the tests I added here.

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:192]: ACK

File: gslib/tests/test_mv.py
Comment (dilipped) [reply gslib/tests/test_mv.py:156]: > Not sure what you mean here. E2E tests are normally higher level than making sure underlying functions get called.

You mentioned that test_naming is taking care of testing the behavior. I am saying that `test_naming` is testing the naming behavior, but how do we ensure that mv is actually using those naming behavior. We do not have any unittests that check if naming module is getting called.

> Sure, but I'd rather describe what's happening with a code comment than adding ls API calls and breaking the pattern of the rest of the file.

Regarding the API call, `self.AssertNObjectsInBucket(bucket2_uri, 2)` itself makes the `ls` call, so I don't think you are increasing the number of API calls. Also, I don't think we are breaking any pattern, instead we would be enhancing the quality of the test. The key thing that this test is looking for is the behavior of trying to move multiple directories with same file underneath. For this test, I think the naming of the destination file is important and hence it should be tested. That's the whole point of this PR, to avoid naming conflict. In the current state, all we are testing is that the file is getting moved over. There is nothing here that tells me that the naming was handled correctly for mv command specifically.

> The mv tests are largely duplicates of functionality tested in test_cp.py. The original authors probably meant for this to be lighter weight because (1) E2E tests are costly to run and 

Like I said above, checking for count involves an ls call anyway. So I don't think it was light weight in the first place.

> (2) test_cp.py already does more detailed output analysis.

Right. And if we want to rely on that behavior, we should really ensure that mv calls the cp behavior i.e have some unittests that check if copy_helper etc gets called. Basically if A depends on B, and we have tested B, we can be sure that A is tested by ensuring that A is calling B correctly. But in this case, we do not test if mv is calling cp. So there are two ways to test this, either via E2E test or via unittest test and ensuring that correct calls were made to cp. We have chosen to test this via E2E test, and if we are doing that, we should ensure that the end result is correct and not rely on inner knowledge that mv relies on cp.

> I technically didn't need to add anything to test_mv.py because we already get coverage from test_cp.py, but I wanted to be thorough. It's pushing it too far to break the lightweight testing pattern of the rest of the file, and I'd rather delete all the tests I added here.

I appreciate you writing these tests, but like I said, there is no lightweight testing pattern that we are breaking. The number of API calls remain the same with what I am asking here. 

Regarding deleting all these tests, we need tests for mv because that's what led to having this PR in the first place. We need to ensure the behavior of mv in such cases. Remember that mv had a separate if condition originally which we removed. But, like I said, you need not have e2e tests if you can ensure using unittests that mv will call the cp behavior. That is what would actually make the tests lightweight.

All this is my opinion of course, and feel free to bring in a second reviewer. I am happy to approve this if others feel what I am proposing is not really needed.

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:752]: Done

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:156]: Lightweight as in amount of code and smaller scope. Sorry, I meant to remove the part about using extra resources b/c extra API calls are not being made.

But I think it's definitely breaking a pattern if we start adding `ls` output analysis in a file where all similar tests do not have it (with one exception `test_stdin_args`). I'm not a fan of rewriting redundant code to what I wrote for `test_cp.py`, especially if it breaks a pattern.

Comment (NickGoog) [reply gslib/tests/test_mv.py:156]: > I think we should replace this test with a unit test that simply checks that the appropriate methods/functions are called

That sounds like `test_naming` unless you mean mocking functions and calling `assert_called_with(...)` (I don't suggest that b/c it's usually better practice to test a behavior rather than an underlying mechanism). I can try converting these to `test_naming` unit tests if you want.

> There is not really a good way for me to look at all these tests in mv and say with confidence that this behavior has been already tested. In such cases, the test here should ensure that the behavior is tested.

The behavior being tested is that `mv`'ing 2 files results in 2 files, not 1, which would indicate improper merging on name duplicate. I can add more clarification in code if you want.

> there is not really a reason to follow a pattern IMO

I think the pattern's purpose is to keep tests simple and short, which can be done b/c the more complex behavior is tested in `test_cp.py`.

> I don't think adding 3-4 lines of code is going to make the test heavy.

We have to combine the "count objects in bucket" logic from `AssertNObjectsInBucket` with string matching on the output paths. It will be more than 4 lines but not a lot if we implement a `assert_paths` parameter for `AssertNObjectsInBucket`. But this is unnecessary if you want the `test_mv` tests converted to `test_naming` unit tests (there are other `mv` tests in `test_naming`).

File: gslib/tests/test_naming.py
Comment (NickGoog) [reply gslib/tests/test_naming.py]: `./` is two characters. I'll change it to `relative_path_string`.

Comment (NickGoog) [reply gslib/tests/test_naming.py]: Done, good call

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py]: Fixed

File: gslib/utils/copy_helper.py
Comment (NickGoog) [reply gslib/utils/copy_helper.py:728]: Technically you could change it to:
```
preserve_src_top_level_dirs = ('**' not in src_url.versionless_url_string and (
                                                   has_multiple_top_level_srcs or
                                                   (src_url_is_valid_parent and have_existing_dest_subdir)))
```

But this seemed needlessly less robust and harder to reason about when I wrote it.

Comment (dilipped) [reply gslib/utils/copy_helper.py:728]: I meant dropping `src_url_is_valid_parent` altogether. Do we get the same result without it?

File: gslib/tests/test_mv.py
Comment (dilipped) [reply gslib/tests/test_mv.py:156]: I was suggesting `assert_called_ with`. The reason being, you mentioned that `test_cp` is already testing for the paths. So if that is the case and if we truly want to avoid the redundancy to make it lightweight, we can simply have a test that checks if mv is calling cp behavior. My point being, there is nothing here or in other tests for mv that tells me that mv calls copy_helper or other naming helpers as expected.  This is the knowledge that you and I have, but it does not reflect in the test code. Of course we can add a comment explaining this, but the more self documenting code we have, the better.

> it's usually better practice to test a behavior rather than an underlying mechanism

Testing the behavior is better, but you have to ensure that you are testing it correctly. If you say that you will partially test this because it is already tested by a dependent module, you kinda have to test that you are calling that module correctly.

But anyway, I don't want to block this PR for this. I will leave it for the second reviewer or other experts to break the tie as I feel we both have been repeating the same points and it's not leading anywhere. Will approve the PR once the other open comment is resolved.

File: gslib/utils/copy_helper.py
Comment (dilipped) [reply gslib/utils/copy_helper.py:728]: Nevermind, I think it is required, else you will see .. in the destination path for single top level source.

**Review by dilipped — APPROVED** [review]
There is one open question related to the tests where we need an opinion from another reviewer. Otherwise, looks good.

File: gslib/tests/test_mv.py
Comment (NickGoog) [reply gslib/tests/test_mv.py:156]: As per my last comment, I'm fine trying to convert the `test_mv.py` E2E tests to `test_naming.py` unit tests but just wanted to check if you have objections first

Comment (dilipped) [reply gslib/tests/test_mv.py:156]: I was not suggesting converting this to `test_naming`, but instead adding a unit test in test_mv itself that verifies that naming or copy_helper is called as expected. Adding this to `test_naming` is fine, but I think we would again have the same issue, i.e nothing in mv tests that we are correctly using the "naming" behavior.

Comment (dilipped) [reply gslib/tests/test_mv.py:156]: Also, this is a shortcoming in test_mv itself. I feel in general there should be a test that ensures it calls copy_helper or the cp functionality correctly. So it's not specific to the test you have here.

Comment (NickGoog) [reply gslib/tests/test_mv.py]: Done, left one test as E2E b/c the unit test framework was having trouble with simulating downloads following the `test_naming.py` pattern. The one left as E2E does check output paths.

Comment (dilipped) [reply gslib/tests/test_mv.py]: LGTM

Comment (dilipped) [reply gslib/tests/test_mv.py]: Also, thanks for adding the tests!

**Pull request merged**
